### PR TITLE
Profile Page

### PR DESCRIPTION
**Pull request by devstm**
User Profile Page
Route '/:id'
- refactor user profile page
- add new types for profile page
- add main service name to user Information Card
- and style user Information Card style
- change user profile route to contain user id
- pass reviews data to Reviews Component as props
- add face book and Instagram link to user Info Card and add link to whats app chat
- create work-list Component
- move WorkListProps type to types file
- create function to get user profile information
- create function to get works per page data
- read home top ten reviews from context
localhost_3000_user_2

### CODE REVIEW

File: client/src/Components/WorksList/index.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import React from 'react';
+import { Pagination, Empty } from 'antd';
+import WorkCard from '../WorkCard/WorkCard';
```
Comment (RandSohail) [inline client/src/Components/WorksList/index.tsx:3]: rename workCard.tsx to index.tsx 
and call him directly 
```
import WorkCard from '../WorkCard';
```

Comment (RandSohail) [reply client/src/Components/WorksList/index.tsx:3]: it was fixed in main branch just pull it

File: client/src/Components/WorksList/index.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  worksData, page, handlePageChange, resultCount, isAuth, isLoading, error,
+}:WorkListProps) {
+  return (
+    <div className="works-container">
+      {isLoading ? <SpinierComponent /> : iff(
+        !error,
+        worksData.length ? (
+          <div className="work-card-container">
+            { worksData?.map((work: OnWork) => (
+              <WorkCard key={work?.id} work={work} actions={act} isAuth={isAuth} />
```
Comment (RandSohail) [inline client/src/Components/WorksList/index.tsx:24]: workCard does not need an actions props anymore just pull the main

File: client/src/Pages/Profile.tsx [hunk 3]
```diff
@@ -1,62 +1,92 @@
...
 import {
-  Works, Request, AllWorks, OnWork, request,
+  UserInfoCard, SpinierComponent, WorkList, Reviews, ErrorComponent,
+} from '../Components';
+import {
+  Works, User, TopTenReviews, ProfileDataProps, WorksProps,
 } from '../utils';
+import { UserContext } from '../Context/LoggedUserContext';
+import getUserProfileData from '../Controllers/getProfileData';
+import getWorksData from '../Controllers/getWorks';
```
Comment (RandSohail) [inline client/src/Pages/Profile.tsx:11]: make an index file in the controller's folder and import getWorksData, getProfileData from it directly

File: client/src/Controllers/getProfileData.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  try {
+    setPage(1);
+    setIsLoading(true);
+    const response : ProfileInfoResponse = await request('get', `/providers/${id}`);
+    const { data } : {data: UserResponse} = response;
+    const { user } : {user: User} = data;
+    const { works }: { works: Works } = data;
+    const { reviews }: { reviews: TopTenReviews[] } = data;
+    const { totalReviews }: { totalReviews: number} = data;
+    const { count } = data;
```
Comment (LinaYahya) [inline client/src/Controllers/getProfileData.ts:23]:
```suggestion
interface ProviderDataType {
user: User,
works: Works,
reviews: ,
}
const {user, works, reviews, totalReviews, count}: ProviderDataType = data
```

File: client/src/Controllers/getProfileData.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    successCB(user, works, reviews, totalReviews, count);
+    setIsLoading(false);
```
Comment (LinaYahya) [inline client/src/Controllers/getProfileData.ts:25]:
```suggestion
    setIsLoading(false);
```
Remove this line immediately when the response come back

File: client/src/Pages/Profile.tsx [hunk 6]
```diff
@@ -1,58 +1,91 @@
...
+    setResultCount(count);
+  };
+  const onFailed = (err:any) => {
+    setError(err?.data.msg);
+  };
+  const getUserProfileDataParams: ProfileDataProps = {
+    setPage, setIsLoading, successCB: onSuccess, failedCB: onFailed, id: +id,
+  };
+  const onGetWorkSuccess = (works: Works) => {
+    setWorksData(works);
```
Comment (LinaYahya) [inline client/src/Pages/Profile.tsx:50]: Let us assume that we are in the first page, Storing thw work data for the first page in the state, And now I want to go to the third page,
SO we're setting the whole state with the new works, And now I want to go to the first page again, So I should send the request again while I had the data before,
This doesn't make any sense
I think the best way to do that is to store the new works beside the one we have related to the page contain,

I think this the best way to store the state
```js
const [works, setWorks] = useState({1: worksForPage1, 3: worksForPage3})
```

And then In the use effect, I should check if I have `works[page]`, then no need to send the request

File: server/controllers/user/profile.js [hunk 7]
```diff
@@ -44,12 +49,13 @@ const getProfileInfo = async (req, res, next) => {
...
       reviews: reviewsAndWorks[0],
-      works: reviewsAndWorks[1],
-      test: reviewsAndWorks[2],
+      works: reviewsAndWorks[1].rows,
+      count: reviewsAndWorks[1].count,
       totalReviews: totalReviews / count,
     };
     res.status(200).json({ msg: "Profile user's information", data });
   } catch (error) {
+    console.log(error);
```
Comment (LinaYahya) [inline server/controllers/user/profile.js:58]:
```suggestion
```
Add a console.log in the `serverError` controller

**Review by LinaYahya — CHANGES_REQUESTED** [review]
Good Job Saleh, appreciated effort 👏🏼 
additional small notes to fix,
Just make sure to remove the work when delete is success,
And It's a good to redirect user to not found user or render a customized component for `not found user`, In case of a validation error, when sending a request for `user/aaa` or when user not found as `user/1000`
And keep a comment for rand to add `add Work` button in the provider card

File: client/src/Components/WorksList/index.tsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+                updateWorks={updateWorks}
+                deletedWork={deletedWork}
+                key={work?.id}
+                work={work}
+                isAuth={isAuth}
+              />
+            ))}
+          </div>
+        ) : (
+          <Empty description={(<span>no-data</span>)} />
```
Comment (LinaYahya) [inline client/src/Components/WorksList/index.tsx:32]:
```suggestion
          <Empty description={(<span>{t('no-data')}</span>)} />
```

**Review by LinaYahya — CHANGES_REQUESTED** [review]
Great Job Saleh,
small things to fix
1- add `form.resetFields();` to `handleAddWork` at `workModal` component
2- don't display `review` and `contact` buttons, when `isAuth`
3- try to handle add review, to render the new review in the slider
4- replace `cover` image at `workCard/index` to use Antd image, to get used of preview feature, and change alt to work title

**Review by LinaYahya — APPROVED** [review]

**Review by AbdallahGot — APPROVED** [review]

**Pull request merged**